src/resolve/target.rs

```diff
@@ -0,0 +1,331 @@
+            | (ContextType::LocalRelative, TargetInner::LocalRelative(_))
+            | (ContextType::LocalRelative, TargetInner::Remote(_)) => (),
+
+            (ContextType::LocalAbsolute, TargetInner::Implementation(_)) => (),
+            (ContextType::LocalAbsolute, TargetInner::LocalRelative(_))
+            | (ContextType::LocalAbsolute, TargetInner::Remote(_)) => {
+                meta.diagnostics
+                    .error("permission denied")
+                    .with_error_section(meta.range, "trying to include this")
+                    .note(
+                        "local absolute path not allowed to access remote or local relative files",
+                    )
+                    .emit();
+                return Err(Error::Diagnostic)
+            },
+
+            // TODO: discuss proper remote rules
+            // check CORS
+            (ContextType::Remote, TargetInner::Remote(url)) if meta.context.url.domain() == url.domain() => (),
+            (ContextType::Remote, TargetInner::Remote(_)) => {
+                meta.diagnostics
+                    .error("permission denied")
+                    .with_error_section(meta.range, "trying to include this")
+                    .error("CORS request detected")
+                    .note("remote inclusions can only include remote content from the same domain")
+                    .emit();
+                return Err(Error::Diagnostic)
+            },
+            (ContextType::Remote, _) => {
+                meta.diagnostics
+                    .error("permission denied")
+                    .with_error_section(meta.range, "trying to include this")
+                    .note("remote file can only include other remote content")
+                    .emit();
+                return Err(Error::Diagnostic)
+            },
+
+            (_, TargetInner::LocalAbsolute(path)) => {
+                if !meta.permissions.is_allowed_absolute(path) {
+                    meta.diagnostics
+                        .error("permission denied")
+                        .with_error_section(meta.range, "trying to include this")
+                        .note(format!("not allowed to access absolute path {:?}", path))
+                        .emit();
+                    return Err(Error::Diagnostic)
+                }
+            },
+        }
+        Ok(TargetChecked {
+            inner,
+            meta,
+        })
+    }
+}
+
+impl<'a, 'd> TargetChecked<'a, 'd> {
+    pub fn into_include(self, remote: &Remote) -> Result<Include> {
+        let TargetChecked { inner, meta } = self;
+        match inner {
+            TargetInner::Implementation(command) => {
+                match Command::from_str(&command) {
+                    Ok(command) => Ok(Include::Command(command)),
+                    Err(()) => {
+                        meta.diagnostics
+                            .error(format!(
+                                "no heradoc implementation found for domain {:?}",
```

Comment:
We are mapping domains to target here so I'd rewrite this in terms of the target.

```suggestion
                                "the heradoc domain {:?} does not name a valid command"
```